src/connector/src/sink/mqtt.rs

```diff
@@ -153,22 +223,90 @@ impl MqttSinkWriter {
+                .and_then(|field| {
+                    schema
+                        .fields()
+                        .iter()
+                        .enumerate()
+                        .find(|(_, f)| f.name == field)
+                        .map(|(pos, f)| {
+                            path.push(pos);
+                            &f.data_type
+                        })
+                })
+                .and_then(|dt| {
+                    iter.try_fold(dt, |dt, field| match dt {
+                        DataType::Struct(st) => {
+                            st.iter().enumerate().find(|(_, (s, _))| *s == field).map(
+                                |(pos, (_, dt))| {
+                                    path.push(pos);
+                                    dt
+                                },
+                            )
+                        }
+                        _ => None,
+                    })
+                });
+            if !matches!(dt, Some(DataType::Varchar)) {
+                return Err(SinkError::Config(anyhow!(
+                    "topic field must be of type string but got {:?}",
+                    dt
+                )));
+            }
+        }
+
+        let timestamptz_mode = TimestamptzHandlingMode::from_options(&format_desc.options)?;
 
+        let encoder = match format_desc.format {
+            SinkFormat::AppendOnly => match format_desc.encode {
+                SinkEncode::Json => RowEncoderWrapper::Json(JsonEncoder::new(
+                    schema.clone(),
+                    None,
+                    DateHandlingMode::FromCe,
+                    TimestampHandlingMode::Milli,
+                    timestamptz_mode,
+                    TimeHandlingMode::Milli,
+                )),
+                SinkEncode::Protobuf => {
+                    let (descriptor, sid) = crate::schema::protobuf::fetch_descriptor(
+                        &format_desc.options,
+                        &config.common.topic,
+                        None,
+                    )
+                    .await
+                    .map_err(|e| SinkError::Config(anyhow!(e)))?;
+                    let header = match sid {
+                        None => ProtoHeader::None,
+                        Some(sid) => ProtoHeader::ConfluentSchemaRegistry(sid),
+                    };
+                    RowEncoderWrapper::Proto(ProtoEncoder::new(
+                        schema.clone(),
+                        None,
+                        descriptor,
+                        header,
+                    )?)
+                }
+                _ => {
+                    return Err(SinkError::Config(anyhow!(
+                        "mqtt sink encode unsupported: {:?}",
+                        format_desc.encode,
+                    )))
+                }
+            },
+            _ => {
+                return Err(SinkError::Config(anyhow!(
+                    "Mqtt sink only support append-only mode"
```

Comment:
IIUC, the `new` will be called when streaming job being scheduled, not when running `CREATE SINK`. Have you tried to run a `CREATE SINK` in non-append-only mode? Can that statement be rejected correctly? (May add this in test cases as well.)

Reply:
I added back the `r#type` field and it's validated during the validation phase`